### PR TITLE
Add batch support for various image_ops

### PR DESCRIPTION
**Pull request by JoshVarty**
Working on #8926
I used #7369 as a guide for my work here.

I have added batch support for:

- `flip_left_right`
- `flip_up_down`
- `random_flip_left_right`
- `random_flip_up_down`
- `transpose_image`
- `rot90`

I have corrected existing tests in `image_ops_test.py` and introduced a number of new tests based on existing tests for 3D inputs. 

This is my first contribution to this repository and I have tried to follow the [`contributing`](https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md) guidelines. However, running `pylint` on `image_ops_impl.py` and `image_ops_test.py` revealed a number of pre-existing style violations. I've tried to fix the ones relevant to my work but may have missed some.

### CODE REVIEW

**Comment by JoshVarty** [comment]
Hmm, I can successfully run the sanity checks locally via:
`tensorflow/tools/ci_build/ci_build.sh CPU tensorflow/tools/ci_build/ci_sanity.sh`

Does anyone happen to know how I can get more information on why they're failing on CI? The relevant output says:

```
Step 3 : COPY install/*.sh /install/
tensorflow/tools/ci_build/ci_build.sh: line 130: 26052 Terminated              docker build -t ${DOCKER_IMG_NAME} -f "${DOCKERFILE_PATH}" "${DOCKER_CONTEXT_PATH}"
ERROR: docker build failed. Dockerfile is at /var/lib/jenkins/workspace/tensorflow-pull-requests-sanity/tensorflow/tools/ci_build/Dockerfile.cpu
Build was aborted
Aborted by unknown
Unable to get pull request builder trigger!!
Setting status of abcb89bce4ce4b68cf714e1cccbc13d4eb1309b7 to FAILURE with url https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/11268/ and message: 'FAILURE
 '
Using context: Sanity Checks
Finished: ABORTED
```

File: tensorflow/python/ops/image_ops_impl.py [hunk 1]
```diff
@@ -182,8 +182,38 @@ def _CheckAtLeast3DImage(image, require_static=True):
     return []
 
 
-def fix_image_flip_shape(image, result):
-  """Set the shape to 3 dimensional if we don't know anything else.
+def _EnsureTensorIs4D(image):
+  """Converts `image` to a 4-D Tensor if it is not already one
```
Comment (drpngx) [inline tensorflow/python/ops/image_ops_impl.py]: Add period at end of sentence.

Comment (drpngx) [reply tensorflow/python/ops/image_ops_impl.py]: Thank you for that function.

**Review by drpngx — DISMISSED** [review]

File: tensorflow/python/ops/image_ops_test.py [hunk 2]
```diff
@@ -768,6 +789,33 @@ def testRandomFlipLeftRight(self):
...
       self.assertGreaterEqual(count_unflipped, 1)
 
+  def testRandomFlipLeftRightWithBatch(self):
+    x_np = np.array([[[1, 2, 3], [1, 2, 3]], [[1, 2, 3], [1, 2, 3]]],
+                    dtype=np.uint8).reshape([2, 2, 3, 1])
+    y_np = np.array([[[3, 2, 1], [3, 2, 1]], [[3, 2, 1], [3, 2, 1]]],
+                    dtype=np.uint8).reshape([2, 2, 3, 1])
+
+    with self.test_session(use_gpu=True):
+      x_tf = constant_op.constant(x_np, shape=x_np.shape).eval()
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: Can you set a seed here to make sure this isn't going to be flaky?

File: tensorflow/python/ops/image_ops_test.py [hunk 3]
```diff
@@ -768,6 +789,33 @@ def testRandomFlipLeftRight(self):
...
+          current_y_np = y_np[index]
+
+          if current_y_tf[0][0] == 1:
+            self.assertAllEqual(current_y_tf, current_x_tf)
+            count_unflipped += 1
+          else:
+            self.assertAllEqual(current_y_tf, current_y_np)
+            count_flipped += 1
+      self.assertGreaterEqual(count_flipped, 1)
+      self.assertGreaterEqual(count_unflipped, 1)
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: The number should be binomially distributed around 50, so the stddev of the normal approximation is 5. We should be pretty safe to assert that count_flipped > 20 and count_unflipped > 20. That's still 6 sigma.

File: tensorflow/python/ops/image_ops_test.py [hunk 4]
```diff
@@ -768,6 +789,33 @@ def testRandomFlipLeftRight(self):
...
+
   def testIdempotentUpDown(self):
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: I see now that you're not the one introducing the idempotent issue, can you fix it here too?

File: tensorflow/python/ops/image_ops_test.py [hunk 5]
```diff
@@ -807,6 +877,32 @@ def testRandomFlipUpDown(self):
...
+          current_y_tf = y_tf[index]
+          current_y_np = y_np[index]
+        if current_y_tf[0][0] == 1:
+          self.assertAllEqual(current_y_tf, current_x_tf)
+          count_unflipped += 1
+        else:
+          self.assertAllEqual(current_y_tf, current_y_np)
+          count_flipped += 1
+      self.assertGreaterEqual(count_flipped, 1)
+      self.assertGreaterEqual(count_unflipped, 1)
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: Same comment, let's assert >20, and add a seed.

File: tensorflow/python/ops/image_ops_test.py [hunk 6]
```diff
@@ -807,6 +877,32 @@ def testRandomFlipUpDown(self):
...
+
   def testIdempotentTranspose(self):
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_test.py]: Argh idempotent.

**Review by martinwicke — CHANGES_REQUESTED** [review]
Looks great in general. I have some comments of medium importance. The biggest is: Can you refactor the code so that the common functionality is not repeated?

**Comment by JoshVarty** [comment]
Thanks for taking a look guys. @martinwicke Let me know if you spot any other opportunities for cleanup/refactoring.

File: tensorflow/python/ops/image_ops_impl.py [hunk 7]
```diff
@@ -211,6 +211,45 @@ def _EnsureTensorIs4D(image):
...
   return (image, is_batch)
 
+def _flip_image(image, axis, random=False, seed=None):
+  """
+  Flips image(s) around a given axis.
+
+  Args:
+    image: 4-D Tensor of shape `[batch, height, width, channels]` or
+           3-D Tensor of shape `[height, width, channels]`.
+
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py]: I think you need to get rid of this blank line for the comment to render properly

File: tensorflow/python/ops/image_ops_impl.py [hunk 8]
```diff
@@ -211,6 +211,45 @@ def _EnsureTensorIs4D(image):
...
+    axis:  A Tensor. Must be one of the following types: int32, int64. 1-D.
+           The indices of the dimensions to reverse. Must be in the range
+           [-rank(tensor), rank(tensor))
+
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py]: You'll need to document all arguments.

File: tensorflow/python/ops/image_ops_impl.py [hunk 9]
```diff
@@ -211,6 +211,45 @@ def _EnsureTensorIs4D(image):
...
 
+def _flip_image(image, axis, random=False, seed=None):
+  """
+  Flips image(s) around a given axis.
+
+  Args:
+    image: 4-D Tensor of shape `[batch, height, width, channels]` or
+           3-D Tensor of shape `[height, width, channels]`.
+
+    axis:  A Tensor. Must be one of the following types: int32, int64. 1-D.
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py]: Why 1D? This should be a scalar? 

I understand you can pass 1D to reverse, but unless you use that, I would keep it a scalar here, and pass [axis] to reverse. 

If you want to keep it 1D, you should change the docstring to explain that each components needs to be in the range given, and possibly they have to be ordered? Not sure. Definitely adds complexity you don't really need.

**Review by martinwicke — CHANGES_REQUESTED** [review]
Awesome. So much less code. Thank you! Just minor things in the docstring (and one thing about the arguments to the private function).

File: tensorflow/python/ops/image_ops_impl.py (hunk 9 above)
Comment (JoshVarty) [reply tensorflow/python/ops/image_ops_impl.py]: I agree that it should be a scalar.

One strange thing about this refactor is that it forces the caller to provide an axis relative to a 4-D Tensor, even if they've passed in a 3-D image.

For example `random_flip_up_down()` needs to provide an axis, but doesn't know whether or not the image it's working with is 3-D or a 4-D batch of images. Currently I just pass in `1` with the knowledge that a single image will be extended to 4 dimensions.

File: tensorflow/python/ops/image_ops_impl.py [hunk 10]
```diff
@@ -182,8 +182,81 @@ def _CheckAtLeast3DImage(image, require_static=True):
...
+  return (image, is_batch)
+
+def _flip_image(image, axis, random=False, seed=None):
+  """
+  Flips image(s) around a given axis.
+
+  Args:
+    image:  4-D Tensor of shape `[batch, height, width, channels]` or
+            3-D Tensor of shape `[height, width, channels]`.
+    axis:   A Python integer representing the axis on which the image(s)
```
Comment (JoshVarty) [inline tensorflow/python/ops/image_ops_impl.py]: As mentioned [here](https://github.com/tensorflow/tensorflow/pull/14854#discussion_r155128239):

One strange thing about this refactor is that it forces the caller to provide an axis relative to a 4-D Tensor, even if they've passed in a 3-D image.

For example `random_flip_up_down()` needs to provide an axis, but doesn't know whether or not the image it's working with is 3-D or a 4-D batch of images. Currently I just pass in `1` with the knowledge that a single image will be extended to 4 dimensions.

Comment (martinwicke) [reply tensorflow/python/ops/image_ops_impl.py]: That's a good point -- but I think that's ok. It's made clear in the argument description, and in the end, this is a private utility function, so I'm not too worried about it.

**Review by martinwicke — APPROVED** [review]

File: tensorflow/python/ops/image_ops_impl.py (hunk 10 above)
Comment (martinwicke) [reply tensorflow/python/ops/image_ops_impl.py]: I guess if you wanted to, you could make a symbolic argument: direction='UP_DOWN' or 'LEFT_RIGHT' and interpret that. But again, the function is private, I don't think it's worth it.

Comment (JoshVarty) [reply tensorflow/python/ops/image_ops_impl.py]: 👍

**Comment by jhseu** [comment]
Hey @JoshVarty, thanks for the contribution. Unfortunately, we have to revert this change because it turns out to be a bottleneck in our image models. There wasn't a reasonable way to fix the performance, because it's caused by both the conversion to 4-D and switching to tf.where. We'd want the 4-D version to perform reasonably well before we leave it in.

I added benchmarks here:
https://github.com/tensorflow/tensorflow/pull/15348

You can run it with `bazel run -c opt test_name -- --benchmarks=FlipImageBenchmark`

**Comment by JoshVarty** [comment]
Thanks for the heads up guys. Presumably the other places where 3-D tensors are expanded to 4-D tensors would also be slow, correct? (Perhaps they're not used in your image models so it wasn't noticed until now)

It sounds like two fixes are needed:
1. We shouldn't always expand 3-D tensors to 4-D.
2. We need an alternative to `tf.where` when operating on 4-D tensors.

The first issue I think I know how to fix.

Do you guys have any hunches on how performance for 4-D batches might be fixed?  Is there a performant alternative to `tf.where` I should experiment with? Would a new op for 4-D tensors make sense here?

**Comment by jhseu** [comment]
Yeah, possibly other 3-D to 4-D changes are slower too, but this change was really noticeable in our internal benchmarks.

The `tf.where` issue is because we're flipping all images. With `tf.cond` on a 3-D Tensor, we only flip images as needed.

**Comment by JoshVarty** [comment]
Would you guys be comfortable with a solution that used `tf.cond` when operating on 3-D tensors, but `tf.where` when operating on 4-D tensors? Or is it best to avoid `tf.where` altogether?

**Comment by jhseu** [comment]
I'm comfortable with anything where the recently added benchmarks perform about the same before and after, including with 4-D tensors.

The issue is that we'd discourage users from using 4-D if it's significantly slower than 3-D on a per-image benchmark.

**Comment by jhseu** [comment]
(Note the benchmark didn't include 4-D tests, but it's trivial to add).

**Pull request merged**